modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java

```diff
@@ -252,23 +283,27 @@ private void initConfigurationResolver() {
 
     if (isConfigurationSpecified()) {
       // the config is dynamic
-      configurationResolver = event -> {
-        ConfigurationInstance instance = configurationProvider.get().get(event);
-        if (instance == null) {
-          throw new IllegalModelDefinitionException(format(
-                                                           "Root component '%s' contains a reference to config '%s' but it doesn't exists",
-                                                           getLocation().getRootContainerName(),
-                                                           configurationProvider));
-        }
-
-        return of(instance);
-      };
+      // the optional should be always present at this point because we already checked dependsOnEvent
+      configurationResolver = event -> resolveConfigFromProvider(getConfigurationProvider().get(), event);
     } else {
       // obtain implicit instance
       configurationResolver = event -> extensionManager.getConfiguration(extensionModel, componentModel, event);
     }
   }
 
+  private Optional<ConfigurationInstance> resolveConfigFromProvider(ConfigurationProvider configurationProvider,
```

Comment:
This one was born as a refactor to avoid repeating the logic for the resolution of the `ConfigurationInstance` from a `ConfigurationProvider`.
We used to have something like this:
```
configurationResolver = event -> {
        ConfigurationInstance instance = configurationProvider.get().get(event);
        if (instance == null) {
          throw new IllegalModelDefinitionException(format(
                                                           "Root component '%s' contains a reference to config '%s' but it doesn't exists",
                                                           getLocation().getRootContainerName(),
                                                           configurationProvider));
        }

        return of(instance);
      };
```
Where `configurationResolver` is of type `Function<CoreEvent, Optional<ConfigurationInstance>>`.
Now we just do this:
```
configurationResolver = event -> resolveConfigFromProvider(getConfigurationProvider().get(), event);
```
It seems the `configurationResolver` (which existed before this PR) needs to return an `Optional<ConfigurationInstance>` mostly because the configuration may be optional for the `ExtensionComponent`.
That being said, you are right in that we could have this new method return a ConfigurationInstance and just wrap on the lambda, like this:
```
configurationResolver = event -> of(resolveConfigFromProvider(getConfigurationProvider().get(), event));
```
But we would have to add a similar wrapping a few lines above that one, in the lambda for the dynamic `ValueResolver<ConfigurationProvider>` case. We can definitely do that if you think it will be clearer. I'm not completely decided for one or the other.